altapay.php

```diff
@@ -0,0 +1,2353 @@
+                $orderLines = $this->createOrderStatusOrderLines($amountToCapture);
+                $api->captureAmount($paymentID, $orderLines, $amountToCapture);
+            }
+        } catch (Exception $e) {
+            $cookie = $this->context->cookie;
+
+            // Saves the error in a cookie, to display it if a HTTP redirect occurs:
+            $msg = $e->getMessage();
+            $cookie->altapayError = Tools::displayError('Error trying to change the order status:' . $msg);
+
+            // Saves the error in errors[], to display it if there is no HTTP redirect:
+            $this->context->controller->errors[] = $cookie->altapayError;
+
+            // Saves the error in the database. The function is loaded from helpers file.
+            saveLastErrorMessage($paymentID, $cookie->altapayError);
+        }
+    }
+
+    /**
+     * Method is being triggered when release action is being performed
+     * @param $paymentID
+     * @param $params
+     * @param bool $captureRemainedAmount
+     * @return mixed
+     */
+    public function performRelease($paymentID, $params, $captureRemainedAmount = true)
+    {
+        try {
+            $api = new MerchantAPI();
+            $api->init($this->getAltapayUrl(), $this->getAPIUsername(), $this->getAPIPassword());
+            $paymentDetails = $api->getPaymentDetails($paymentID);
+            $orderReservedAmount = $paymentDetails->getReservedAmount();
+            $orderCapturedAmount = $paymentDetails->getCapturedAmount();
+            $refundedAmount = $paymentDetails->getRefundedAmount();
+
+            if ($orderCapturedAmount == 0 && $refundedAmount == 0) {
+                $releaseResult = $api->release($paymentID);
+                if ($releaseResult->wasSuccessful()) {
+                    // Saves the payment stays in the database. The function is loaded from helpers file.
+                    updatePaymentStatus($paymentID, "Payment Released");
+                }
+            } else if ($orderCapturedAmount == $refundedAmount && $refundedAmount == $orderReservedAmount || $refundedAmount == $orderReservedAmount) {
+                $api->release($paymentID);
+
+            } else {
+                $api->release($paymentID);
+            }
+        } catch (Exception $e) {
+            $cookie = $this->context->cookie;
+
+            // Saves the error in a cookie, to display it if a HTTP redirect occurs:
+            $msg = json_decode($e->getMessage());
+            $cookie->altapayError = Tools::displayError('Error trying to change the order status: ' . $msg->responseMsg);
+
+            // Saves the error in errors[], to display it if there is no HTTP redirect:
+            $this->context->controller->errors[] = $cookie->altapayError;
+
+            // Saves the error in the database. The function is loaded from helpers file.
+            saveLastErrorMessage($paymentID, $cookie->altapayError);
+
+            return $cookie->altapayError;
+        }
```

Comment:
This block appears to be repeated in multiple function, maybe it would be a good candidate for a helper method.